### PR TITLE
Amend I/O reform RFC to address issues with flush

### PR DESCRIPTION
**Pull request by mzabaluev**

### CODE REVIEW

**Comment by glaebhoerl** [comment]
[rendered](https://github.com/mzabaluev/rust-rfcs/blob/io-write-tweaks/text/0517-io-os-reform.md)

File: text/0517-io-os-reform.md [hunk 1]
```diff
@@ -596,13 +616,17 @@ method, clients of `Read` can easily recover the `push` method.
 ### `Write`
 [Write]: #write
 
-The `Writer` trait is cut down to even smaller size:
+The `Writer` trait is cut down to even smaller size, and `flush` is taken
```
Comment (glaebhoerl) [inline text/0517-io-os-reform.md]: `Writer` doesn't really get any smaller - all of the same methods are still there - just refactored.

Comment (jmesmon) [inline text/0517-io-os-reform.md]: note that this is an edit of a previous RFC which removed quite a few methods from `Write`. The statement you're commenting was preserved from that one.

File: text/0517-io-os-reform.md [hunk 2]
```diff
@@ -251,6 +252,25 @@ In the meantime, this RFC proposes to phase out the use of impls for
...
+  bytestream output. Other output traits are likely needed for specific data
+  types, such as [TextWriter][rfc272] for UTF-8 text, and types implementing
+  those traits may need a flush operation as well, intrinsically implemented
+  or possibly a trait method. There should be a way to provide this operation
+  without duplicating it in different output traits or forcing writer-like
+  types to implement byte output.
+
+* A convention on `into_inner` methods on composite writers and `BufStream`
+  is to flush the underlying writer before returning it. The deep flushing
+  may be undesirable if the consumer wants to continue writing.
```
Comment (alexcrichton) [inline text/0517-io-os-reform.md]: After taking another look at the implementations, I don't think that this "deep flushing" is happening (at least not in `std::io`), perhaps this was just a problem in `old_io`?

**Comment by alexcrichton** [comment]
After thinking about this a bit today, I'm not quite sure that we may want the split. I agree that other abstractions may want a `flush` method as well, but I'm not sure that this incarnation of `Flush` is appropriate for all use cases. My biggest concern is that the usage of `io::Error` may make it inadequate for the general use case. For example `std::fmt::Write` uses a different error type than I/O. Additionally, delegation of a `flush` method from one trait to another (if the `flush` method is duplicated among traits) arguably takes fewer lines to express than an explicit implementation of `Flush` itself.

I feel like there may be a place for a generic `Flush`, but I don't find the motivation compelling enough to try to introduce the general trait at this time.

**Comment by alexcrichton** [comment]
cc @aturon

**Comment by kballard** [comment]
I'm not sure what the use-case of a generic `Flush` is. I can't think of any reason why a function or type would want to be able to `flush()` a stream without first having written something to it, and if it knows how to write to the stream then it obviously knows what type the stream is (and therefore is already bound on something like `Write`).

I also don't see what the benefit is of moving the declaration of `fn flush(&mut self) -> io::Result<()>` out of the `Write` trait and into a separate trait. Given the hypothetical `TextWriter`, having both `Write` and `TextWriter` declare `fn flush(&mut self) -> io::Result<()>` is not code duplication in any meaningful sense. Any implementation of these traits would need to implement `flush()` regardless of where it's declared, and the fact that both `Write` and TextWriter` may want to have an identical method `fn flush(&mut self) -> io::Result<()>` does not necessarily mean that this is in fact a truly generic flush method that is appropriate for all stream types.

In fact, the only benefit that comes to mind is a type that implements both `Write` and this hypothetical `TextWriter` would only need to implement `flush()` once, instead of twice. But that seems like an edge case that's not worth worrying about, and it's also not necessarily obvious that the implementation would even want to be the same in both cases (I can think of some esoteric stream types that might want the ability to have flushing behavior that differs from `Write::flush`).

**Comment by mzabaluev** [comment]
@alexcrichton:

> For example `std::fmt::Write` uses a different error type than I/O.

This is actually an issue of its own: `std::fmt::Error` should expose `std::io::Error` as a possible cause.

File: text/0517-io-os-reform.md (hunk 2 above)
Comment (mzabaluev) [inline text/0517-io-os-reform.md]: Indeed, it's just that the documentation is vague.

**Comment by mzabaluev** [comment]
I have filed issue rust-lang/rust#23386 to correct the documentation of `into_inner`, and #978 to amend the I/O RFC accordingly.
As the generic `Flush` trait did not get much support, I'm closing this PR.

**Pull request closed without merging**